File: scripts/sync_shalom.py
#!/usr/bin/env python3
"""
=============================================================================
PIPELINE DE INGESTA Y SINCRONIZACIÓN SHALOM PERÚ (PYTHON 3)
=============================================================================
Descarga el 100% del catálogo oficial de agencias de Shalom Perú mediante
paginación dinámica continua (sin cortes de límite implícito) y realiza un
Batch Upsert masivo en PostgreSQL / Supabase con PostGIS y Ubigeo.
"""

import os
import sys
import json
import time
import urllib.request
import urllib.error

# Forzar UTF-8 en consola de Windows
if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")
if hasattr(sys.stderr, "reconfigure"):
    sys.stderr.reconfigure(encoding="utf-8")

# Configuración de Entorno
API_KEY = os.environ.get("SHALOM_API_KEY", "sk_qm4rm5ivepety4ausqnubkfegp4yr2lnqu3p4q55oc3v4yzw3oma")
BASE_URL = os.environ.get("SHALOM_API_URL", "https://api.shalom-api-peru.com").rstrip("/")
SUPABASE_URL = os.environ.get("VITE_SUPABASE_URL", "https://uwmdjsxwetjvsxsdngko.supabase.co").rstrip("/")
SUPABASE_KEY = os.environ.get("VITE_SUPABASE_ANON_KEY", "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InV3bWRqc3h3ZXRqdnN4c2RuZ2tvIiwicm9sZSI6ImFub24iLCJpYXQiOjE3ODY2NDE5MTEsImV4cCI6MjEwMjIxNzkxMX0.KaqryIyoe4IDQGTJD_cswZkW-wfgnMcyV9tJoWxHMq8")


def normalize_agency(raw: dict) -> dict:
    """
    Normaliza un registro de agencia extrayendo:
    ID, Nombre Completo, Dirección, Departamento, Provincia, Distrito, Ubigeo y Coordenadas.
    """
    # Horario
    horario_raw = raw.get("horario")
    schedule = "Lunes a Sábado: 8:00 AM - 8:00 PM"
    if isinstance(horario_raw, str) and horario_raw.strip():
        schedule = horario_raw.strip()
    elif isinstance(horario_raw, dict) and horario_raw.get("hora_atencion"):
        schedule = horario_raw["hora_atencion"].strip()

    # Ubicación geográfica administrativa
    dep = str(raw.get("departamento") or "LIMA").upper().strip()
    prov = str(raw.get("provincia") or "LIMA").upper().strip()
    dist = str(raw.get("distrito") or raw.get("lugar_over") or raw.get("zona") or "CENTRO").upper().strip()
    code = raw.get("abrebiatura") or raw.get("code") or None
    name = str(raw.get("nombre") or "").strip()
    address = str(raw.get("direccion") or "").strip()

    # Ubigeo
    ubi_id = raw.get("ubi_id") or raw.get("ubigeo")
    ubigeo = str(ubi_id).zfill(6) if ubi_id else None

    # Nombre completo normalizado
    code_tag = f" (CÓDIGO: {code})" if code else ""
    full_name = f"{dep} / {prov} / {dist} / {name}{f' - {address}' if address else ''}{code_tag}".upper()

    return {
        "id": raw["id"],
        "code": code,
        "name": name,
        "full_name": full_name,
        "department": dep,
        "province": prov,
        "district": dist,
        "ubigeo": ubigeo,
        "dep_id": raw.get("dep_id"),
        "prov_id": raw.get("prov_id"),
        "dist_id": raw.get("dist_id"),
        "address": address,
        "phone": raw.get("telefono") or "(01) 500-7878",
        "schedule": schedule,
        "latitude": raw.get("latitud"),
        "longitude": raw.get("longitud"),
        "is_active": True,
        "updated_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
    }


def fetch_all_agencies_from_api(per_page: int = 50, delay_seconds: float = 0.5) -> list:
    """
    Descarga el 100% de agencias iterando página tras página hasta que la API devuelva [].
    """
    print("=" * 60)
    print("🚀 INICIANDO DESCARGA DINÁMICA DE AGENCIAS SHALOM PERÚ (PYTHON)")
    print("=" * 60)
    print(f"📡 API Endpoint: {BASE_URL}/v1/agencies")
    print(f"🔑 API Key: {'Configurada' if API_KEY else 'No encontrada'}")
    print("-" * 60)

    all_raw_agencies = []
    page = 1
    has_more = True

    while has_more:
        url = f"{BASE_URL}/v1/agencies?page={page}&per_page={per_page}"
        print(f"📥 Solicitando Página {page}...")

        req = urllib.request.Request(
            url,
            headers={
                "X-API-Key": API_KEY,
                "User-Agent": "Shalom-Python-Sync/1.0",
                "Accept": "application/json"
            }
        )

        try:
            with urllib.request.urlopen(req, timeout=15) as response:
                if response.status != 200:
                    print(f"❌ Error HTTP {response.status} en página {page}")
                    break
                data = json.loads(response.read().decode("utf-8"))

                # Manejar array directo, data o items
                items = data if isinstance(data, list) else (data.get("items") or data.get("data") or [])

                if not items:
                    print(f"🏁 No hay más agencias en la página {page}. Finalizando bucle.")
                    has_more = False
                else:
                    all_raw_agencies.extend(items)
                    print(f"  ✅ Página {page} OK: +{len(items)} agencias (Acumulado: {len(all_raw_agencies)})")

                    if len(items) < per_page:
                        print(f"🏁 Última página alcanzada ({len(items)} < {per_page}).")
                        has_more = False
                    else:
                        page += 1
                        time.sleep(delay_seconds)  # Rate Limit Safe: 60 req/min

        except Exception as e:
            print(f"❌ Error al consultar página {page}: {e}")
            break

    print("-" * 60)
    print(f"🎉 Descarga HTTP completada. Total agencias recibidas: {len(all_raw_agencies)}")
    return all_raw_agencies


def fetch_existing_agencies_db() -> dict:
    """
    Obtiene el estado previo de agencias en Supabase para cálculo de Diff.
    """
    if not SUPABASE_URL or not SUPABASE_KEY:
        return {}

    url = f"{SUPABASE_URL}/rest/v1/shalom_agencies?select=id,name,address,schedule,is_active"
    req = urllib.request.Request(
        url,
        headers={
            "apikey": SUPABASE_KEY,
            "Authorization": f"Bearer {SUPABASE_KEY}",
            "Accept": "application/json"
        }
    )
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            if resp.status == 200:
                rows = json.loads(resp.read().decode("utf-8"))
                return {int(r["id"]): r for r in rows}
    except Exception as e:
        print(f"⚠️ No se pudo consultar estado previo en DB: {e}")
    return {}


def batch_upsert_supabase(normalized_agencies: list, batch_size: int = 100):
    """
    Inserta o actualiza por lotes (Batch Upsert) directamente en Supabase PostgreSQL.
    """
    if not SUPABASE_URL or not SUPABASE_KEY:
        print("⚠️ Supabase no configurado en variables de entorno. Saltando carga SQL.")
        return

    print(f"💾 Guardando {len(normalized_agencies)} agencias en Supabase por lotes de {batch_size}...")

    upsert_url = f"{SUPABASE_URL}/rest/v1/shalom_agencies"
    headers = {
        "apikey": SUPABASE_KEY,
        "Authorization": f"Bearer {SUPABASE_KEY}",
        "Content-Type": "application/json",
        "Prefer": "resolution=merge-duplicates"
    }

    total_inserted = 0
    for i in range(0, len(normalized_agencies), batch_size):
        batch = normalized_agencies[i:i + batch_size]
        payload = json.dumps(batch).encode("utf-8")

        req = urllib.request.Request(upsert_url, data=payload, headers=headers, method="POST")

        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                if resp.status in (200, 201):
                    total_inserted += len(batch)
                    print(f"  💾 Lote {i // batch_size + 1} guardado ({total_inserted}/{len(normalized_agencies)})...")
                else:
                    print(f"  ⚠️ Respuesta inesperada en lote {i // batch_size + 1}: {resp.status}")
        except urllib.error.HTTPError as e:
            err_body = e.read().decode("utf-8", errors="ignore")
            print(f"  ❌ Error HTTP al insertar lote {i // batch_size + 1}: {e.code} - {err_body}")
        except Exception as e:
            print(f"  ❌ Error de conexión al insertar lote: {e}")

    print(f"✅ Batch Upsert finalizado: {total_inserted} registros confirmados en base de datos.")


def deactivate_closed_agencies(deactivate_ids: list):
    """
    Marca como is_active=False las agencias que Shalom dio de baja en su API.
    """
    if not deactivate_ids or not SUPABASE_URL or not SUPABASE_KEY:
        return

    print(f"🛑 Desactivando {len(deactivate_ids)} agencias cerradas por Shalom...")
    ids_param = ",".join(str(i) for i in deactivate_ids)
    url = f"{SUPABASE_URL}/rest/v1/shalom_agencies?id=in.({ids_param})"
    payload = json.dumps({"is_active": False, "updated_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())}).encode("utf-8")

    req = urllib.request.Request(
        url,
        data=payload,
        headers={
            "apikey": SUPABASE_KEY,
            "Authorization": f"Bearer {SUPABASE_KEY}",
            "Content-Type": "application/json",
        },
        method="PATCH"
    )
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            if resp.status in (200, 204):
                print(f"  ✅ {len(deactivate_ids)} agencias marcadas como inactivas en DB.")
    except Exception as e:
        print(f"  ⚠️ Error desactivando agencias: {e}")


def main():
    start_time = time.time()
    
    # 1. Ingesta completa dinámica
    raw_agencies = fetch_all_agencies_from_api(per_page=50, delay_seconds=0.5)

    if not raw_agencies:
        print("❌ No se pudieron obtener agencias de la API.")
        sys.exit(1)

    # 2. Normalización de campos
    normalized = [normalize_agency(item) for item in raw_agencies]
    api_ids = {int(a["id"]) for a in normalized}

    # 3. Detección diferencial contra base de datos
    existing_db = fetch_existing_agencies_db()
    new_count = 0
    updated_count = 0
    reactivated_count = 0
    to_deactivate_ids = []

    for item in normalized:
        item_id = int(item["id"])
        if item_id not in existing_db:
            new_count += 1
        else:
            prev = existing_db[item_id]
            if prev.get("is_active") is False:
                reactivated_count += 1
            elif (
                prev.get("name") != item["name"] or
                prev.get("address") != item["address"] or
                prev.get("schedule") != item["schedule"]
            ):
                updated_count += 1

    for db_id, prev in existing_db.items():
        if prev.get("is_active") is not False and db_id not in api_ids:
            to_deactivate_ids.append(db_id)

    print("=" * 60)
    print("📊 DIFERENCIAL DETECTADO:")
    print(f"  ✨ Nuevas agencias detectadas       : {new_count}")
    print(f"  📝 Agencias con cambios (nombre/dir): {updated_count}")
    print(f"  🔄 Agencias reactivadas             : {reactivated_count}")
    print(f"  🛑 Agencias cerradas/eliminadas     : {len(to_deactivate_ids)}")
    print("=" * 60)

    # 4. Guardar snapshot JSON local y actualizar fallback TS
    os.makedirs("data", exist_ok=True)
    with open("data/shalom_agencies_full.json", "w", encoding="utf-8") as f:
        json.dump(normalized, f, ensure_ascii=False, indent=2)
    print("📁 Respaldo JSON guardado en: data/shalom_agencies_full.json")

    # 5. Inserción Masiva en BD
    batch_upsert_supabase(normalized, batch_size=100)

    # 6. Desactivar bajas
    if to_deactivate_ids:
        deactivate_closed_agencies(to_deactivate_ids)

    # 7. Estadísticas y Departamentos detectados
    departments = {}
    for a in normalized:
        d = a["department"]
        departments[d] = departments.get(d, 0) + 1

    print("=" * 60)
    print("📊 RESUMEN POR DEPARTAMENTOS:")
    for dep, count in sorted(departments.items()):
        print(f"  • {dep.ljust(20)}: {count} agencias")
    
    elapsed = round(time.time() - start_time, 2)
    print("=" * 60)
    print(f"🏁 SINCRONIZACIÓN EXITOSA EN {elapsed}s | TOTAL ACTIVAS: {len(normalized)}")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: scripts/sync-shalom.mjs
import { createClient } from '@supabase/supabase-js';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// 1. Cargar variables de entorno desde .env
const envPath = path.resolve(__dirname, '../.env');
const envVars = {};
if (fs.existsSync(envPath)) {
  const content = fs.readFileSync(envPath, 'utf8');
  content.split('\n').forEach((line) => {
    const trimmed = line.trim();
    if (trimmed && !trimmed.startsWith('#') && trimmed.includes('=')) {
      const [key, ...rest] = trimmed.split('=');
      envVars[key.trim()] = rest.join('=').trim();
    }
  });
}

const SUPABASE_URL = process.env.VITE_SUPABASE_URL || envVars.VITE_SUPABASE_URL || 'http://api.89.117.73.97.sslip.io';
const SUPABASE_KEY = process.env.VITE_SUPABASE_ANON_KEY || envVars.VITE_SUPABASE_ANON_KEY || 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InV3bWRqc3h3ZXRqdnN4c2RuZ2tvIiwicm9sZSI6ImFub24iLCJpYXQiOjE3ODY2NDE5MTEsImV4cCI6MjEwMjIxNzkxMX0.KaqryIyoe4IDQGTJD_cswZkW-wfgnMcyV9tJoWxHMq8';
const SHALOM_API_KEY = process.env.SHALOM_API_KEY || envVars.SHALOM_API_KEY || 'sk_qm4rm5ivepety4ausqnubkfegp4yr2lnqu3p4q55oc3v4yzw3oma';
const SHALOM_API_URL = (process.env.SHALOM_API_URL || envVars.SHALOM_API_URL || 'https://api.shalom-api-peru.com').replace(/\/+$/, '');

const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

function normalizeItem(raw) {
  let scheduleText = 'Lunes a Sábado: 8:00 AM - 8:00 PM';
  if (typeof raw.horario === 'string' && raw.horario.trim()) {
    scheduleText = raw.horario.trim();
  } else if (raw.horario && typeof raw.horario === 'object' && raw.horario.hora_atencion) {
    scheduleText = raw.horario.hora_atencion.trim();
  }

  const department = (raw.departamento || 'LIMA').toUpperCase().trim();
  const province = (raw.provincia || 'LIMA').toUpperCase().trim();
  const district = (raw.distrito || raw.lugar_over || raw.zona || 'CENTRO').toUpperCase().trim();
  const code = raw.abrebiatura || raw.code || null;
  const name = (raw.nombre || '').trim();
  const address = (raw.direccion || '').trim();

  // Ubigeo (6 dígitos)
  let ubigeo = null;
  if (raw.ubi_id) {
    ubigeo = String(raw.ubi_id).padStart(6, '0');
  } else if (raw.ubigeo) {
    ubigeo = String(raw.ubigeo).padStart(6, '0');
  }

  // Nombre Completo Detallado Normalizado
  const codeTag = code ? ` (CÓDIGO: ${code})` : '';
  const fullName = `${department} / ${province} / ${district} / ${name}${address ? ` - ${address}` : ''}${codeTag}`.toUpperCase();

  return {
    id: raw.id,
    code,
    name,
    nombre: name,
    full_name: fullName,
    department,
    departamento: department,
    province,
    provincia: province,
    district,
    distrito: district,
    ubigeo,
    dep_id: typeof raw.dep_id === 'number' ? raw.dep_id : null,
    prov_id: typeof raw.prov_id === 'number' ? raw.prov_id : null,
    dist_id: typeof raw.dist_id === 'number' ? raw.dist_id : null,
    address,
    direccion: address,
    phone: raw.telefono || '(01) 500-7878',
    telefono: raw.telefono || '(01) 500-7878',
    schedule: scheduleText,
    horario: scheduleText,
    latitude: typeof raw.latitud === 'number' ? raw.latitud : null,
    longitude: typeof raw.longitud === 'number' ? raw.longitud : null,
    is_active: true,
    updated_at: new Date().toISOString()
  };
}

function toDbRow(item) {
  return {
    id: item.id,
    code: item.code,
    name: item.name,
    full_name: item.full_name,
    department: item.department,
    province: item.province,
    district: item.district,
    ubigeo: item.ubigeo,
    dep_id: item.dep_id,
    prov_id: item.prov_id,
    dist_id: item.dist_id,
    address: item.address,
    phone: item.phone,
    schedule: item.schedule,
    latitude: item.latitude,
    longitude: item.longitude,
    is_active: item.is_active,
    updated_at: item.updated_at
  };
}

async function runShalomSync() {
  console.log('====================================================');
  console.log('🚀 INICIANDO SINCRONIZACIÓN SHALOM PERÚ (POSTGIS DB)');
  console.log('====================================================');
  console.log(`📡 URL API: ${SHALOM_API_URL}`);
  console.log(`🔑 Key detectada: ${SHALOM_API_KEY ? 'Sí (sk_...)' : 'No'}`);
  console.log(`🗄️  Supabase URL: ${SUPABASE_URL}`);
  console.log('----------------------------------------------------');

  const startTime = Date.now();
  let allAgencies = [];
  let page = 1;
  const perPage = 50;
  let hasMore = true;

  console.log('🔄 Iniciando ingesta dinámica de agencias Shalom...');

  while (hasMore) {
    const url = `${SHALOM_API_URL}/v1/agencies?page=${page}&per_page=${perPage}`;
    console.log(`📥 Solicitando Página ${page}...`);

    try {
      const res = await fetch(url, {
        headers: {
          'X-API-Key': SHALOM_API_KEY,
          'User-Agent': 'Incomi-Cron-Sync/1.0'
        }
      });

      if (!res.ok) {
        console.error(`❌ Error en Página ${page}: ${res.status} ${res.statusText}`);
        break;
      }

      const json = await res.json();
      // Extraer array sin importar si viene en items, data o directamente como arreglo
      const items = Array.isArray(json) ? json : (json.items || json.data || []);

      if (!items || items.length === 0) {
        console.log(`🏁 No hay más agencias en la página ${page}. Finalizando bucle.`);
        hasMore = false;
      } else {
        allAgencies = [...allAgencies, ...items];
        console.log(`  ✅ Página ${page} procesada: +${items.length} agencias (Acumulado: ${allAgencies.length})`);

        // Si devolvió menos elementos que perPage, alcanzamos la última página
        if (items.length < perPage) {
          hasMore = false;
        } else {
          page++;
          // Pausa de 500ms para no superar el Rate Limit (60 req/min)
          await new Promise((resolve) => setTimeout(resolve, 500));
        }
      }
    } catch (err) {
      console.error(`❌ Error de red en página ${page}:`, err);
      break;
    }
  }

  console.log('----------------------------------------------------');
  console.log(`🎉 Ingesta HTTP finalizada. Total agencias recibidas de Shalom: ${allAgencies.length}`);
  
  const normalized = allAgencies.map(normalizeItem);
  const apiIdsSet = new Set(normalized.map(a => Number(a.id)));

  // 1. Obtener agencias existentes en Supabase para cálculo de Diff
  console.log('🔍 Analizando estado previo en Supabase para detección de cambios...');
  const { data: existingRows } = await supabase
    .from('shalom_agencies')
    .select('id, name, address, schedule, is_active');

  const existingMap = new Map((existingRows || []).map(r => [Number(r.id), r]));
  
  let newCount = 0;
  let updatedCount = 0;
  let reactivatedCount = 0;
  const toDeactivateIds = [];

  normalized.forEach(item => {
    const existing = existingMap.get(Number(item.id));
    if (!existing) {
      newCount++;
    } else {
      if (existing.is_active === false) {
        reactivatedCount++;
      } else if (
        existing.name !== item.name ||
        existing.address !== item.address ||
        existing.schedule !== item.schedule
      ) {
        updatedCount++;
      }
    }
  });

  // Detectar agencias que Shalom eliminó del API oficial
  existingMap.forEach((existing, dbId) => {
    if (existing.is_active !== false && !apiIdsSet.has(dbId)) {
      toDeactivateIds.push(dbId);
    }
  });

  console.log('----------------------------------------------------');
  console.log(`📊 DIFERENCIAL DETECTADO:`);
  console.log(`  ✨ Nuevas agencias detectadas       : ${newCount}`);
  console.log(`  📝 Agencias con cambios (nombre/dir): ${updatedCount}`);
  console.log(`  🔄 Agencias reactivadas             : ${reactivatedCount}`);
  console.log(`  🛑 Agencias cerradas/eliminadas     : ${toDeactivateIds.length}`);
  console.log('----------------------------------------------------');

  // 2. Guardar / Batch Upsert en Supabase (Lotes de 100)
  console.log('💾 Procediendo a Batch Upsert en Supabase PostgreSQL...');
  let upserted = 0;
  const batchSize = 100;
  for (let i = 0; i < normalized.length; i += batchSize) {
    const batch = normalized.slice(i, i + batchSize).map(toDbRow);
    const { error } = await supabase
      .from('shalom_agencies')
      .upsert(batch, { onConflict: 'id' });

    if (error) {
      console.error(`❌ Error en lote ${Math.floor(i / batchSize) + 1}:`, error.message);
    } else {
      upserted += batch.length;
      console.log(`  💾 Lote ${Math.floor(i / batchSize) + 1} sincronizado (${upserted}/${normalized.length})...`);
    }
  }

  // 3. Desactivar agencias que ya no existen en la API
  if (toDeactivateIds.length > 0) {
    console.log(`🛑 Desactivando ${toDeactivateIds.length} agencias cerradas por Shalom...`);
    const { error: deactError } = await supabase
      .from('shalom_agencies')
      .update({ is_active: false, updated_at: new Date().toISOString() })
      .in('id', toDeactivateIds);

    if (deactError) {
      console.warn('⚠️ Error al desactivar agencias:', deactError.message);
    } else {
      console.log(`  ✅ ${toDeactivateIds.length} agencias marcadas como inactivas.`);
    }
  }

  // 4. Actualizar respaldo local en JSON y código estático
  try {
    fs.mkdirSync('data', { recursive: true });
    fs.writeFileSync('data/shalom_agencies_full.json', JSON.stringify(normalized, null, 2), 'utf8');

    const staticTs = `import { ShalomAgency } from '../types/database.types';

export const DEPARTAMENTOS_PERU = [
  'TODOS',
  'AMAZONAS', 'ANCASH', 'APURIMAC', 'AREQUIPA', 'AYACUCHO', 'CAJAMARCA',
  'CALLAO', 'CUSCO', 'HUANCAVELICA', 'HUANUCO', 'ICA', 'JUNIN',
  'LA LIBERTAD', 'LAMBAYEQUE', 'LIMA', 'LORETO', 'MADRE DE DIOS', 'MOQUEGUA',
  'PASCO', 'PIURA', 'PUNO', 'SAN MARTIN', 'TACNA', 'TUMBES', 'UCAYALI'
];

export const SHALOM_AGENCIES: ShalomAgency[] = ${JSON.stringify(normalized, null, 2)};
`;
    fs.writeFileSync('src/data/shalomAgencies.ts', staticTs, 'utf8');
    console.log('📁 Respaldo offline sincronizado en data/ y src/data/shalomAgencies.ts');
  } catch (fsErr) {
    console.warn('⚠️ No se pudo actualizar el archivo local:', fsErr.message);
  }

  // 5. Prueba rápida de consulta espacial PostGIS
  console.log('🧪 Verificando consulta espacial PostGIS (Agencias cercanas a Lima Centro)...');
  const testLat = -12.046374;
  const testLng = -77.042793;

  const { data: nearby, error: rpcError } = await supabase.rpc('get_nearby_shalom_agencies', {
    user_lat: testLat,
    user_lng: testLng,
    max_limit: 3
  });

  if (rpcError) {
    console.warn('⚠️ Nota al probar RPC espacial:', rpcError.message);
  } else if (nearby && nearby.length > 0) {
    console.log('📍 3 Agencias más cercanas encontradas:');
    nearby.forEach((ag, idx) => {
      console.log(`  ${idx + 1}. [${ag.code || 'N/A'}] ${ag.name} (${(ag.distance_meters / 1000).toFixed(2)} km)`);
    });
  }

  const durationSec = ((Date.now() - startTime) / 1000).toFixed(1);
  console.log('====================================================');
  console.log(`🏁 Sincronización completada exitosamente en ${durationSec}s.`);
  console.log(`📊 TOTAL ACTIVAS: ${normalized.length} AGENCIAS`);
  console.log('====================================================');
}

runShalomSync().catch((err) => {
  console.error('❌ Error fatal en sincronización:', err);
  process.exit(1);
});
